rust/cyclotron-core/src/worker.rs

```diff
@@ -267,22 +309,228 @@ impl Worker {
         // lock after the if check, makes the compiler think the lock is held across
         // the await point.
         {
-            let pending = self.pending.lock().unwrap();
+            let pending = self.running.lock().unwrap();
             if !pending.contains_key(&job_id) {
-                return Err(QueueError::UnknownJobId(job_id));
+                return Err(JobError::UnknownJobId(job_id).into());
             }
         }
 
         dead_letter(&self.pool, job_id, reason).await
     }
 
     /// Passing None here will clear the blob
-    pub fn set_blob(&self, job_id: Uuid, blob: Option<Bytes>) -> Result<(), QueueError> {
-        let mut pending = self.pending.lock().unwrap();
+    pub fn set_blob(&self, job_id: Uuid, blob: Option<Bytes>) -> Result<(), JobError> {
+        let mut pending = self.running.lock().unwrap();
         pending
             .get_mut(&job_id)
-            .ok_or(QueueError::UnknownJobId(job_id))?
+            .ok_or(JobError::UnknownJobId(job_id))?
             .blob = Some(blob);
         Ok(())
     }
 }
+
+// Started by each worker on creation, just loops seeing if the passed batch can be flushed, and
+// if it can, flushing it.
+async fn flush_loop(
+    pool: PgPool,
+    batch: Weak<Mutex<FlushBatch>>,
+    max_buffered: usize,
+    max_bytes: usize,
+    interval: Duration,
+) {
+    loop {
+        let Some(batch) = batch.upgrade() else {
+            // The batch has been dropped, we should exit.
+            break;
+        };
+        // Contemplating sync mutexes on the tree of woe.
+        let mut to_flush = { batch.lock().unwrap().take() };
+        if to_flush.should_flush(max_buffered, max_bytes) {
+            if let Err(e) = to_flush.flush(&pool).await {
+                error!("Error flushing batch: {:?}", e);
+            }
+        }
+        // We can always merge the taken batch back into the pending batch - on successful
+        // flush, the taken batch will be empty, and on failure, we need to re-queue those updates.
+        // TRICKY - we take care not to bind the lock here. Compilation WILL fail if it's bound,
+        // because it makes this future !Send, and the tokio::spawn above will fail, but in case
+        // we change the looping strategy, I'm calling it out explicitly too.
+        batch.lock().unwrap().merge(to_flush);
+        tokio::time::sleep(interval.to_std().unwrap()).await;
```

Comment:
Question: How do we "stop" this flush loop? Like on the program exiting?

Reply:
The `Weak::upgrade` on line 342 fails indicating the worker who "owns" the batch (holds a non-weak reference to it) has been dropped.